### PR TITLE
Feature/ref 491 xliff generation automation

### PR DESCRIPTION
**Pull request by ntripathy**

### CODE REVIEW

File: src/start.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                        sourceResources['ng2-template'][sourceId].target = targetResources[targetNameSpace][targetId].target;
+                        translationFound = true;
+                        break;
+                    }
+                }
+                if(!translationFound) {
+                    translationNotFoundList.push(sourceResources['ng2-template'][sourceId].source);
+                }
+            }
+            sourceResult.targetLanguage = "zh-CN";
```
Comment (m9bromu) [inline src/start.js]: Surely this should change for every file? This is not in your LanguageInfo array.  Maybe you should just include the target language token value and you generate all the file names from it?  Its hard to keep things really regular from an array of array of strings.   So then you would just have:
const LANGUAGE_TOKENS = ['de','es',....'zh-CN']

Alternately, you could generate your list of tokens by parsing the files in XLFSRC_DIR if this will ever be done again, possibly with more languages.

File: src/start.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                var translationFound = false;
+                for (targetId in targetResources[targetNameSpace]) {
+                    if (sourceResources['ng2-template'][sourceId].source === targetResources[targetNameSpace][targetId].source) {
+                        sourceResources['ng2-template'][sourceId].target = targetResources[targetNameSpace][targetId].target;
+                        translationFound = true;
+                        break;
+                    }
+                }
+                if(!translationFound) {
+                    translationNotFoundList.push(sourceResources['ng2-template'][sourceId].source);
```
Comment (m9bromu) [inline src/start.js]: Shouldn't you also record the target so you can report which language files don't have translations for a source message?

File: src/start.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                throw err;
+            }
+        });
+    });
+}
+
+function createTranslationNotFoundFile(translationNotFoundList, fileName) {
+    var fs = require('fs');
+    for (i=0; i<translationNotFoundList.length; i++) {
+        fs.appendFile(TEXT_NOT_FOUND_DIR + fileName + '.txt', translationNotFoundList[i] + "\n", 'utf8', (err) => {
```
Comment (m9bromu) [inline src/start.js]: For developers we probably want a list ordered by source strings with maybe a spec of which languages are missing them.
For translators we would want a list ordered by language. 
So put tuples (source, language) into translationNotFoundList and sort them twice.

File: src/start.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        const targetResources = targetResult.resources;
+        xliff12ToJs(sourceData, (sErr, sourceResult) => {
+            const sourceResources = sourceResult.resources;
+            var translationNotFoundList = [];
+            for (sourceId in sourceResources['ng2-template']) {
+                var translationFound = false;
+                for (targetId in targetResources[targetNameSpace]) {
+                    if (sourceResources['ng2-template'][sourceId].source === targetResources[targetNameSpace][targetId].source) {
+                        sourceResources['ng2-template'][sourceId].target = targetResources[targetNameSpace][targetId].target;
+                        translationFound = true;
```
Comment (m9bromu) [inline src/start.js]: What happens if the source or language files have the same source entry listed twice?   Is that in scope for the tool to detect?  (Probably not if this is a single use tool.  Maybe if it will be used again in a month or two.  Ask the UI developers.)

File: README.md [hunk 5]
```diff
@@ -1,2 +1,31 @@
 # rw-xliff-converter
 RefWorks xliff converter
+
+
+## Purpose
+
+The purpose of this module to automate the xliff generation process for different languages which replaces a manual tedious job.
+
+## Prerequisites
+The scripts accepts a set of input file in 'xlfsrc' directory. These language specific files have been generated from the PO files present in the 'src/main/resources/public/i18n/po/' under 'pqd' repo.
```
[the remark this answers is not in the document]
Comment (ntripathy) [reply README.md]: Yes it xlfsrc is a top level directory within the repo.

File: src/start.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+function processLanguageFile(data_source, masterFile, targetNameSpace, outputFile) {
+    var fs = require('fs');
+    fs.readFile(XLF_SRC_DIR + masterFile,'utf8', function(err, data_target) {
+        process(data_source, data_target, targetNameSpace, outputFile);
+    });
+}
+
+function process(sourceData, targetData, targetNameSpace, outputFileName)
+{
+   const xliff12ToJs = require('xliff/xliff12ToJs');
```
[the remark this answers is not in the document]
Comment (ntripathy) [reply src/start.js]: Yes the function "processLanguageFile" and "process"gets called for each language file mentioned in the LANGUAGE_INFO array declared at the begining.

File: src/start.js (hunk 2 above)
Comment (ntripathy) [reply src/start.js]: The current code outputs the text translation not found for each file and are created in the "notfound" folder of the project root directory.  However, it looks same across all the languages. The translation is missing for the same set of text in each languages.

File: src/start.js (hunk 1 above)
Comment (ntripathy) [reply src/start.js]: This has been fixed by reading the target language from source file itself as given below:
ourceResult.targetLanguage = targetResult.targetLanguage;

File: package.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "xliff-proj",
```
[the remark this answers is not in the document]
Comment (ntripathy) [reply package.json]: Fixed.

File: package.json [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  "version": "1.0.0",
+  "description": "",
```
[the remark this answers is not in the document]
Comment (ntripathy) [reply package.json]: Done.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  "dependencies": {
+    "fs": "0.0.1-security",
+    "xliff": "git+https://github.com/locize/xliff.git"
+  },
+  "devDependencies": {},
+  "scripts": {
+    "start": "node src/start.js",
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
```
[the remark this answers is not in the document]
Comment (ntripathy) [reply package.json]: Removed author for now.

File: src/start.js (hunk 2 above)
Comment (m9bromu) [reply src/start.js]: Good enough for the current usage

File: src/start.js (hunk 3 above)
Comment (m9bromu) [reply src/start.js]: Per your comment above, not needed now

**Comment by ntripathy** [comment]
Thanks Joe 👍
